File: Leetcode/21-merge-two-sorted-lists.cpp
class Solution {
    public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        ListNode *head = NULL;
        ListNode *tail = head;
        
        if(list1->val <= list2->val)
        {
            head = list1;
            list1 = list1->next;
        }
        else
        {
            head = list2;
            list2 = list2->next;
        }
        
        while(list1 != NULL && list2 != NULL)
        {    
            if(list1->val <= list2->val)
            {
                tail = new ListNode(list1->val);
                list1 = list1->next;
            }
            else if(list1->val > list2->val)
            {
                tail = new ListNode(list2->val);
                list2 = list2->next;
            }

            ListNode* tail = head;
            while(tail->next)
            {
                tail = tail->next;
            }

            tail->next = tail;
        }
    
        tail = head;
        while(tail->next)
                tail = tail->next;
                
        tail->next = (list1 == NULL) ? list2 : list1;
        
        return head;
    }
};

File: Leetcode/21-Merge-Two-Sorted-Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        left = list1
        right = list2
        
        mergedHead = mergedTail = None
        
        while left and right:
            toAppend = None
            if left.val <= right.val:
                toAppend = left.val
                left = left.next
            else:
                toAppend = right.val
                right = right.next

            if mergedTail:
                mergedTail.next = ListNode(toAppend)
                mergedTail = mergedTail.next
            else:
                mergedHead = ListNode(toAppend)
                mergedTail = mergedHead
        
        while left:
            if mergedTail:
                mergedTail.next = ListNode(left.val)
                mergedTail = mergedTail.next
            else:
                mergedHead = ListNode(left.val)
                mergedTail = mergedHead
            left = left.next
        
        while right:
            if mergedTail:
                mergedTail.next = ListNode(right.val)
                mergedTail = mergedTail.next
            else:
                mergedHead = ListNode(right.val)
                mergedTail = mergedHead
            right = right.next
    
        return mergedHead
            